src/core/lib/security/credentials/xds/xds_credentials.cc

```diff
@@ -61,44 +61,57 @@ bool XdsVerifySubjectAlternativeNames(
-            request->peer_info.san_names.uri_names_size,
-            xds_certificate_provider_->GetSanMatchers(cluster_name_)) &&
-        !XdsVerifySubjectAlternativeNames(
-            request->peer_info.san_names.ip_names,
-            request->peer_info.san_names.ip_names_size,
-            xds_certificate_provider_->GetSanMatchers(cluster_name_)) &&
-        !XdsVerifySubjectAlternativeNames(
-            request->peer_info.san_names.dns_names,
-            request->peer_info.san_names.dns_names_size,
-            xds_certificate_provider_->GetSanMatchers(cluster_name_))) {
-      *sync_status = absl::Status(
-          absl::StatusCode::kUnauthenticated,
-          "SANs from certificate did not match SANs from xDS control plane");
-    }
-    return true; /* synchronous check */
+const char kTlsCertificateVerifierTypeXds[] = "xds";
+
+}  // namespace
+
+//
+// XdsCertificateVerifier
+//
+
+XdsCertificateVerifier::XdsCertificateVerifier(
+    RefCountedPtr<XdsCertificateProvider> xds_certificate_provider,
+    std::string cluster_name)
+    : grpc_tls_certificate_verifier(kTlsCertificateVerifierTypeXds),
+      xds_certificate_provider_(std::move(xds_certificate_provider)),
+      cluster_name_(std::move(cluster_name)) {}
+
+bool XdsCertificateVerifier::Verify(
+    grpc_tls_custom_verification_check_request* request,
+    std::function<void(absl::Status)>, absl::Status* sync_status) {
+  GPR_ASSERT(request != nullptr);
+  if (!XdsVerifySubjectAlternativeNames(
+          request->peer_info.san_names.uri_names,
+          request->peer_info.san_names.uri_names_size,
+          xds_certificate_provider_->GetSanMatchers(cluster_name_)) &&
+      !XdsVerifySubjectAlternativeNames(
+          request->peer_info.san_names.ip_names,
+          request->peer_info.san_names.ip_names_size,
+          xds_certificate_provider_->GetSanMatchers(cluster_name_)) &&
+      !XdsVerifySubjectAlternativeNames(
+          request->peer_info.san_names.dns_names,
+          request->peer_info.san_names.dns_names_size,
+          xds_certificate_provider_->GetSanMatchers(cluster_name_))) {
+    *sync_status = absl::Status(
+        absl::StatusCode::kUnauthenticated,
+        "SANs from certificate did not match SANs from xDS control plane");
   }
-  void Cancel(grpc_tls_custom_verification_check_request*) override {}
+  return true; /* synchronous check */
+}
 
- private:
-  RefCountedPtr<XdsCertificateProvider> xds_certificate_provider_;
-  std::string cluster_name_;
-};
+void XdsCertificateVerifier::Cancel(
+    grpc_tls_custom_verification_check_request*) {}
 
-}  // namespace
+int XdsCertificateVerifier::cmp_impl(
+    const grpc_tls_certificate_verifier* other) const {
+  auto* o = static_cast<const XdsCertificateVerifier*>(other);
+  if (xds_certificate_provider_ == o->xds_certificate_provider_ &&
+      cluster_name_ == o->cluster_name_) {
+    return 0;
+  }
+  return QsortCompare(static_cast<const grpc_tls_certificate_verifier*>(this),
```

Comment:
I don't think we need to fall back on comparing the pointer values if we're comparing all of the data members anyway.  Instead, I think we can do something like this:

```
int r = QsortCompare(xds_certificate_provider_, o->xds_certificate_provider_);
if (r != 0) return r;
return cluster_name_.compare(o->cluster_name_);
```